### PR TITLE
Fix TransformReplay::getMatchedLeafPosWithoutReplay*

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I am tunning `TransformReplay::getMatchedLeafPosWithoutReplay*`, which checks if replay can be skipped so that I can `TORCH_CHECK(skip_replay)` in my `InlinePropagator`. This will completely separate the concept of transform replay and setting compute-at position.

What I observe is, sometimes, even immediately after `TransformReplay::{replayPasC,replayCasP}`, the skip check still says we need to replay. This is not the correct behavior.

In order to make the just-replayed tensor passes the skip check, we need to make sure that the producer's reduction IDs in a CasP replay are skipped. This matches with the behavior of `TransformReplay::{replayPasC,replayCasP}`. Note that reduction IDs in a PasC replay should not be skipped, because a PasC replay will push the reductions to the end of the domain.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
~This causes new failures in my compute at PR, but I believe this PR is doing the correct thing, and this should be merged without worrying about my compute at PR.~
Edit: fixed

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 1]
```diff
@@ -24079,6 +24079,72 @@ TEST_F(NVFuserTest, FusionIssue1785Repro_CUDA) {
   testValidate(&fusion, cg_outputs, {in1, in2}, {tv_ref}, __LINE__, __FILE__);
 }
 
+TEST_F(NVFuserTest, FusionSkipReplay_CUDA) {
+  struct TransformPropagatorWithCheck : public TransformPropagator {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:24083]: Do we want to make this a global class in our test cases, and replace all `TransformPropagator`s with `TransformPropagatorWithCheck`? If we want to do this, it will be a separate PR.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:24083]: I don't see the harm in doing so, if you think it'll help us catch issues in the future, sure!

**Review by csarofeen — APPROVED** [review]
LGTM
